<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Home ¬∑ ArDCA</title><link rel="canonical" href="https://pagnani.github.io/ArDCA/"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit">ArDCA</span></div><form class="docs-search" action="search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>Home</a><ul class="internal"><li><a class="tocitem" href="#Package-Features"><span>Package Features</span></a></li><li><a class="tocitem" href="#Overview"><span>Overview</span></a></li><li class="toplevel"><a class="tocitem" href="#Usage"><span>Usage</span></a></li><li class="toplevel"><a class="tocitem" href="#Multithreading"><span>Multithreading</span></a></li><li class="toplevel"><a class="tocitem" href="#Tutorial"><span>Tutorial</span></a></li><li><a class="tocitem" href="#Load-ArDCA-package"><span>Load ArDCA package</span></a></li><li><a class="tocitem" href="#Learn-the-autoregressive-parameters"><span>Learn the autoregressive parameters</span></a></li><li><a class="tocitem" href="#.-Sequence-Generation"><span>1. Sequence Generation</span></a></li><li><a class="tocitem" href="#.-Contact-Prediction"><span>2. Contact Prediction</span></a></li><li><a class="tocitem" href="#.-Predicting-mutational-effects"><span>3. Predicting mutational effects</span></a></li><li><a class="tocitem" href="#index"><span>Methods Reference</span></a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Home</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Home</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/pagnani/ArDCA/blob/master/docs/src/index.md#L" title="Edit on GitHub"><span class="docs-icon fab">ÔÇõ</span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="ArDCA"><a class="docs-heading-anchor" href="#ArDCA">ArDCA</a><a id="ArDCA-1"></a><a class="docs-heading-anchor-permalink" href="#ArDCA" title="Permalink"></a></h1><ul><li>Autoreressive model learning for protein inference.</li></ul><h2 id="Package-Features"><a class="docs-heading-anchor" href="#Package-Features">Package Features</a><a id="Package-Features-1"></a><a class="docs-heading-anchor-permalink" href="#Package-Features" title="Permalink"></a></h2><ul><li>Learn model from multiple sequence alignment</li><li>Sample from the model </li></ul><p>See the <a href="#index">Index</a> for the complete list of documented functions and types.</p><h2 id="Overview"><a class="docs-heading-anchor" href="#Overview">Overview</a><a id="Overview-1"></a><a class="docs-heading-anchor-permalink" href="#Overview" title="Permalink"></a></h2><p>Protein families are given in form of multiple sequence alignments (MSA) <span>$D = (a^m_i |i = 1,\dots,L;\,m = 1,\dots,M)$</span> of <span>$M$</span> proteins of aligned length <span>$L$</span>. The entries <span>$a^m_i$</span> equal either one of the standard 20 amino acids, or the alignment gap <span>$‚Äì$</span>. In total, we have <span>$q = 21$</span> possible different symbols in D. The aim of unsupervised generative modeling is to earn a statistical model <span>$P(a_1,\dots,a_L)$</span> of (aligned) full-length sequences, which faithfully reflects the variability found in <span>$D$</span>: sequences belonging to the protein family of interest should have comparably high probabilities, unrelated sequences very small probabilities. Here we propose a computationally efficient approach based on autoregressive models. </p><p>We start from the exact decomposition:</p><p class="math-container">\[P(a_1,\dots,a_L) = P(a_1) \cdot P(a_2|a_1) \cdot \dots \cdot P(a_L|a_1,\dots,a_{L-1})\]</p><p>Here, we use the following parametrization:</p><p class="math-container">\[P(a_i | a_1,\dots,a_{i-1}) = \frac{\exp \left\{ h_i(a_i) + \sum_{j=1}^{i-1} J_{i,j}(a_i,a_j)\right\} }{z_i(a_1,\dots,a_{i-1})}\,,\]</p><p>where:</p><p class="math-container">\[z_i(a_1,\dots,a_{i-1})= \sum_{a=1}^{q} \exp \left\{ h_i(a) + \sum_{j=1}^{i-1} J_{i,j}(a,a_j)\right\} \,,\]</p><p>is a the normalization factor. In machine learning, this parametrization is known as soft-max regression, the generalization of logistic regression to multi-class labels.</p><h1 id="Usage"><a class="docs-heading-anchor" href="#Usage">Usage</a><a id="Usage-1"></a><a class="docs-heading-anchor-permalink" href="#Usage" title="Permalink"></a></h1><p>The typical pipeline to use the package is:</p><ul><li>Compute ArDCA parameters from a multiple sequence alignment:</li></ul><pre><code class="language-none">julia&gt; arnet,arvar=ardca(filefasta; kwds...)</code></pre><ul><li>Generate <code>100</code> new sequences, and store it in an <span>$L\times 100$</span> array of integers.</li></ul><pre><code class="language-none">julia&gt; Zgen =  sample(arnet,100);</code></pre><h1 id="Multithreading"><a class="docs-heading-anchor" href="#Multithreading">Multithreading</a><a id="Multithreading-1"></a><a class="docs-heading-anchor-permalink" href="#Multithreading" title="Permalink"></a></h1><p>To fully exploit the the multicore parallel computation, julia should be invoked with</p><pre><code class="language-none">$ julia -t nthreads # put here nthreads equal to the number of cores you want to use</code></pre><p>If you want to set permanently the number of threads to the desired value, you can either create a default environment variable <code>export JULIA_NUM_THREADS=24</code> in your <code>.bashrc</code>. More information <a href="https://docs.julialang.org/en/v1.6/manual/multi-threading/">here</a></p><h1 id="Tutorial"><a class="docs-heading-anchor" href="#Tutorial">Tutorial</a><a id="Tutorial-1"></a><a class="docs-heading-anchor-permalink" href="#Tutorial" title="Permalink"></a></h1><p>We will assume we have a Multiple Sequence Alignment (MSA)in FASTA format. We aim at</p><ol><li>Given a MSA, generate a sample</li><li>Given a MSA, predict contacts</li><li>Given a MSA, predict the mutational effect in all (ungapped) position of a given target sequence </li></ol><h2 id="Load-ArDCA-package"><a class="docs-heading-anchor" href="#Load-ArDCA-package">Load ArDCA package</a><a id="Load-ArDCA-package-1"></a><a class="docs-heading-anchor-permalink" href="#Load-ArDCA-package" title="Permalink"></a></h2><p>The following cell loads the package <code>ArDCA</code> (<em>Warning</em>: the first time it takes a while)</p><ul><li>The <code>mypkgdir</code> variable should be set to your <code>path/to/package</code> dir.</li></ul><p>We will use the PF00014 protein family available in <code>data/PF14/</code> folder of the package/</p><pre><code class="language-none">mypkgdir = normpath(joinpath(pwd(),&quot;..&quot;))
datadir=joinpath(mypkgdir,&quot;data&quot;) # put here your path
using Pkg
Pkg.activate(mypkgdir)
using ArDCA</code></pre><h2 id="Learn-the-autoregressive-parameters"><a class="docs-heading-anchor" href="#Learn-the-autoregressive-parameters">Learn the autoregressive parameters</a><a id="Learn-the-autoregressive-parameters-1"></a><a class="docs-heading-anchor-permalink" href="#Learn-the-autoregressive-parameters" title="Permalink"></a></h2><p>As a preliminary step, we learn the field and the coupling parameters <span>$h,J$</span> from the MSA. To do so we use the <code>ardca</code> method that return the parameters (stored in <code>arnet</code> in the cell below), and the alignment in numerical format and other algorithms variables (stored in <code>arvar</code> in the cell below). The default autoregressive order is set to <code>:ENTROPIC</code>. We set the <span>$L_2$</span> regularization to 0.02 for the <span>$J$</span> and 0.001 for the <span>$h$</span>.</p><p>The keyword arguments for the <code>ardca</code> method are (with their default value):</p><ul><li><p><code>epsconv::Real=1.0e-5</code> (convergenge parameter)</p></li><li><p><code>maxit::Int=1000</code> (maximum number of iteration - don&#39;t change)</p></li><li><p><code>verbose::Bool=true</code> (set to <code>false</code> to suppress printing on screen)</p></li><li><p><code>method::Symbol=:LD_LBFGS</code> (optimization method)</p></li><li><p><code>permorder::Union{Symbol,Vector{Ti}}=:ENTROPIC</code> (permutation order). Possible values are: <code>[:NATURAL, :ENTROPIC, :REV_ENTROPIC, :RANDOM]</code> or a custom permutation vector.</p></li></ul><pre><code class="language-none">arnet,arvar=ardca(&quot;data/PF14/PF00014_mgap6.fasta.gz&quot;, verbose=false, lambdaJ=0.02,lambdaH=0.001);</code></pre><h2 id=".-Sequence-Generation"><a class="docs-heading-anchor" href="#.-Sequence-Generation">1. Sequence Generation</a><a id=".-Sequence-Generation-1"></a><a class="docs-heading-anchor-permalink" href="#.-Sequence-Generation" title="Permalink"></a></h2><p>We now generate <code>M</code> sequences using the <code>sample</code> method:</p><pre><code class="language-none">M = 1_000;
generated_alignment = sample(arnet,M);</code></pre><p>The generated alignment has is a  ùêø√óùëÄ  matrix of integer where  ùêø  is the alignment&#39;s length, and  ùëÄ  the number of samples.</p><p>Interestingly, we for each sequence we can also compute the likelihood with the sample<em>with</em>weights method.</p><pre><code class="language-none">loglikelihood,generated_alignment = sample_with_weights(arnet,M);</code></pre><h2 id=".-Contact-Prediction"><a class="docs-heading-anchor" href="#.-Contact-Prediction">2. Contact Prediction</a><a id=".-Contact-Prediction-1"></a><a class="docs-heading-anchor-permalink" href="#.-Contact-Prediction" title="Permalink"></a></h2><p>We can compute the epistatic score for residue-residue contact prediction. To do so, we can use the <code>epistatic_score</code> method. The epistatic score is computed on any target sequence of the MSA. Empirically, it turns out the the final score does not depend much on the choice of the target sequence. </p><p>The autput is contained in a <code>Vector</code> of <code>Tuple</code> ranked in descendic score order. Each <code>Tuple</code> contains <span>$i,j,s_{ij}$</span> where <span>$s_{ij}$</span> is the epistatic score of the residue pair <span>$i,j$</span>. The residue numbering is that of the MSA, and not of the unaligned full sequences.</p><pre><code class="language-none">target_sequence = 1
score=epistatic_score(arnet,arvar,target_sequence)</code></pre><h2 id=".-Predicting-mutational-effects"><a class="docs-heading-anchor" href="#.-Predicting-mutational-effects">3. Predicting mutational effects</a><a id=".-Predicting-mutational-effects-1"></a><a class="docs-heading-anchor-permalink" href="#.-Predicting-mutational-effects" title="Permalink"></a></h2><p>For any reference sequence, we can easily predict the mutational effect for all single mutants. Of course we can extract this information only for the <em>non-gapped</em> residues of the target sequence. </p><p>This is done with the <code>dms_single_site</code> method, which returns a <code>q√óL</code> matrix <code>D</code> containing <span>$\log(P(mut))/\log(P(wt))$</span> for all single site mutants of the reference sequence <code>seqid</code> (i.e. the so-called wild type sequence), and <code>idxgap</code> a vector of indices of the residues of the reference sequence that contain gaps (i.e. the 21 amino-acid) for which the score has no sense and is set by convention to <code>+Inf</code>.</p><p>A negative value indicate a beneficial mutation, a value 0 indicate the wild-type amino-acid.</p><pre><code class="language-none">target_sequence = 1
D,idxgap=dms_single_site(arnet,arvar,target_sequence)</code></pre><h2 id="index"><a class="docs-heading-anchor" href="#index">Methods Reference</a><a id="index-1"></a><a class="docs-heading-anchor-permalink" href="#index" title="Permalink"></a></h2><ul><li><a href="#ArDCA.ardca-Tuple{String}"><code>ArDCA.ardca</code></a></li><li><a href="#ArDCA.ardca-Union{Tuple{Ti}, Tuple{Matrix{Ti}, Vector{Float64}}} where Ti&lt;:Integer"><code>ArDCA.ardca</code></a></li><li><a href="#ArDCA.dms_single_site-Tuple{ArNet, ArVar, Int64}"><code>ArDCA.dms_single_site</code></a></li><li><a href="#ArDCA.sample-Tuple{ArNet, Int64}"><code>ArDCA.sample</code></a></li><li><a href="#ArDCA.sample_with_weights-Tuple{ArNet, Any}"><code>ArDCA.sample_with_weights</code></a></li><li><a href="#ArDCA.softmax!-Tuple{Vector{Float64}}"><code>ArDCA.softmax!</code></a></li></ul><article class="docstring"><header><a class="docstring-binding" id="ArDCA.ardca-Tuple{String}" href="#ArDCA.ardca-Tuple{String}"><code>ArDCA.ardca</code></a> ‚Äî <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">ardca(filename::String; kwds...)</code></pre><p>Run <a href="#ArDCA.ardca-Tuple{String}"><code>ardca</code></a> on the fasta alignment in <code>filename</code></p><p><strong>Examples</strong></p><pre><code class="language-none">julia&gt; arnet, arvar =  ardca(&quot;pf14.fasta&quot;, permorder=:ENTROPIC)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/pagnani/ArDCA/blob/8cf5fd4ba5565f9895d033599a05705bd137b0a5/src/ar.jl#LL52-L59">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="ArDCA.ardca-Union{Tuple{Ti}, Tuple{Matrix{Ti}, Vector{Float64}}} where Ti&lt;:Integer" href="#ArDCA.ardca-Union{Tuple{Ti}, Tuple{Matrix{Ti}, Vector{Float64}}} where Ti&lt;:Integer"><code>ArDCA.ardca</code></a> ‚Äî <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">ardca(Z::Array{Ti,2},W::Vector{Float64}; kwds...)</code></pre><p>Auto-regressive analysis on the L√óM alignment <code>Z</code> (numerically encoded in 1,‚Ä¶,21), and the <code>M</code>-dimensional normalized  weight vector <code>W</code>.</p><p>Return two <code>struct</code>: <code>::ArNet</code> (containing the inferred hyperparameters) and <code>::ArVar</code></p><p>Optional arguments:</p><ul><li><code>fracmax::Real=0.3</code> maximum fraction of insert in the sequence</li><li><code>remove_dups::Bool=true</code> if <code>true</code> remove duplicated sequences</li><li><code>theta=:auto</code> if <code>:auto</code> compute reweighint automatically. Otherwise set a <code>Float64</code> value <code>0 ‚â§ theta ‚â§ 1</code></li><li><code>lambdaJ::Real=0.01</code> coupling L‚ÇÇ regularization parameter (lagrange multiplier)</li><li><code>lambdaH::Real=0.01</code> field L‚ÇÇ regularization parameter (lagrange multiplier)</li><li><code>epsconv::Real=1.0e-5</code> convergence value in minimzation</li><li><code>maxit::Int=1000</code> maximum number of iteration in minimization</li><li><code>verbose::Bool=true</code> set to <code>false</code> to stop printing convergence info on <code>stdout</code></li><li><code>method::Symbol=:LD_LBFGS</code> optimization strategy see <a href="https://github.com/JuliaOpt/NLopt.jl"><code>NLopt.jl</code></a> for other options</li><li><code>permorder::Union{Symbol,Vector{Ti}}=:ENTROPIC</code> permutation order. Possible values are <code>:NATURAL,:ENTROPIC,:REV_ENTROPIC,:RANDOM</code> or a custom permutation vector</li></ul><p><strong>Examples</strong></p><pre><code class="language-none">julia&gt; arnet, arvar= ardca(Z,W,lambdaJ=0,lambdaH=0,permorder=:REV_ENTROPIC,epsconv=1e-12);</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/pagnani/ArDCA/blob/8cf5fd4ba5565f9895d033599a05705bd137b0a5/src/ar.jl#LL2-L25">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="ArDCA.dms_single_site-Tuple{ArNet, ArVar, Int64}" href="#ArDCA.dms_single_site-Tuple{ArNet, ArVar, Int64}"><code>ArDCA.dms_single_site</code></a> ‚Äî <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">dms_single_site(arnet::ArNet, arvar::ArVar, seqid::Int; pc::Float64=0.1)</code></pre><p>Return a <code>q√óL</code> matrix of containing <code>-log(P(mut))/log(P(seq))</code> for all single site mutants of the reference sequence <code>seqid</code>, and a vector of the indices of the residues of the reference sequence that contain gaps (i.e. the 21 amino-acid) for which the score has no sense and is set by convention to <code>+Inf</code>. A negative value indicate a beneficial mutation, a value 0 indicate the wild-type amino-acid.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/pagnani/ArDCA/blob/8cf5fd4ba5565f9895d033599a05705bd137b0a5/src/utils.jl#LL211-L220">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="ArDCA.sample-Tuple{ArNet, Int64}" href="#ArDCA.sample-Tuple{ArNet, Int64}"><code>ArDCA.sample</code></a> ‚Äî <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">sample(arnet::ArNet, msamples::Int)</code></pre><p>Return a generated alignment in the form of a <code>N √ó msamples</code>  matrix of type <code>::Matrix{Int}</code>  </p><p><strong>Examples</strong></p><pre><code class="language-none">julia&gt; arnet,arvar=ardca(&quot;file.fasta&quot;,verbose=true,permorder=:ENTROPIC, lambdaJ=0.001,lambdaH=0.001);
julia&gt; Zgen=Zgen=sample(arnet,1000);</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/pagnani/ArDCA/blob/8cf5fd4ba5565f9895d033599a05705bd137b0a5/src/utils.jl#LL97-L105">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="ArDCA.sample_with_weights-Tuple{ArNet, Any}" href="#ArDCA.sample_with_weights-Tuple{ArNet, Any}"><code>ArDCA.sample_with_weights</code></a> ‚Äî <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">sample_with_weights(arnet::ArNet, msamples::Int)</code></pre><p>Return a generated alignment in the form of a <code>N √ó msamples</code>  matrix of type <code>::Matrix{Int}</code> and the relative probabilities under the module</p><p><strong>Examples</strong></p><pre><code class="language-none">julia&gt; arnet,arvar=ardca(&quot;file.fasta&quot;,verbose=true,permorder=:ENTROPIC, lambdaJ=0.001,lambdaH=0.001);
julia&gt; Wgen,Zgen=sample_with_weights(arnet,1000);</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/pagnani/ArDCA/blob/8cf5fd4ba5565f9895d033599a05705bd137b0a5/src/utils.jl#LL133-L142">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="ArDCA.softmax!-Tuple{Vector{Float64}}" href="#ArDCA.softmax!-Tuple{Vector{Float64}}"><code>ArDCA.softmax!</code></a> ‚Äî <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia">softmax(x::AbstractArray{&lt;:Real})</code></pre><p>Return the <a href="https://en.wikipedia.org/wiki/Softmax_function"><code>softmax transformation</code></a> applied to <code>x</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/pagnani/ArDCA/blob/8cf5fd4ba5565f9895d033599a05705bd137b0a5/src/utils.jl#LL90-L93">source</a></section></article></article><nav class="docs-footer"><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Sunday 28 March 2021 12:15">Sunday 28 March 2021</span>. Using Julia version 1.6.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
